stdlib/LinearAlgebra/src/generic.jl

```diff
@@ -1627,6 +1627,70 @@ const NumberArray{T<:Number} = AbstractArray{T}
 exactdiv(a, b) = a/b
 exactdiv(a::Integer, b::Integer) = div(a, b)
 
+# Bareiss algorithm
+function bareiss_update!(zero!, M::Matrix, k, swapto, pivot, prev_pivot)
+    for i in k+1:size(M, 2), j in k+1:size(M, 1)
+        M[j,i] = exactdiv(M[j,i]*pivot - M[j,k]*M[k,i], prev_pivot)
+    end
+    zero!(M, k+1:size(M, 1), k)
+end
+
+function bareiss_update!(zero!, M::AbstractMatrix, k, swapto, pivot, prev_pivot)
+    V = @view M[k+1:end, k+1:end]
+    V .= exactdiv.(V * pivot - M[k+1:end, k] * M[k, k+1:end]', prev_pivot)
+    zero!(M, k+1:size(M, 1), k)
+end
+
+function bareiss_update_virtual_colswap!(zero!, M::AbstractMatrix, k, swapto, pivot, prev_pivot)
+    V = @view M[k+1:end, :]
+    V .= exactdiv.(V * pivot - M[k+1:end, swapto[2]] * M[k, :]', prev_pivot)
+    zero!(M, k+1:size(M, 1), swapto[2])
+end
+
+bareiss_zero!(M, i, j) = M[i,j] .= zero(eltype(M))
+
+function find_pivot_col(M, i)
+    p = findfirst(!iszero, @view M[i,i:end])
+    p === nothing && return nothing
+    idx = CartesianIndex(i, p + i - 1)
+    (idx, M[idx])
+end
+
+function find_pivot_any(M, i)
+    p = findfirst(!iszero, @view M[i:end,i:end])
+    p === nothing && return nothing
+    idx = p + CartesianIndex(i - 1, i - 1)
+    (idx, M[idx])
+end
+
+const bareiss_colswap = (Base.swapcols!, Base.swaprows!, bareiss_update!, bareiss_zero!)
+const bareiss_virtcolswap = ((M,i,j)->nothing, Base.swaprows!, bareiss_update_virtual_colswap!, bareiss_zero!)
+
+"""
+    bareiss!(M)
+
+Perform Bareiss's fraction-free row-reduction algorithm on the matrix `M`.
+Optionally, a specific pivoting method may be specified.
+"""
+function bareiss!(M::AbstractMatrix,
+                     (swapcols!, swaprows!, update!, zero!) = bareiss_colswap;
```

Comment:
I agree that it looks odd to do the deconstruction in the argument list. I think it would be easier to read if the deconstruction happened in the function body. Also, should this argument be documented?